clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -6513,17 +6513,32 @@ static void handleSYCLIntelNoGlobalWorkOffsetAttr(Sema &S, Decl *D,
   S.AddSYCLIntelNoGlobalWorkOffsetAttr(D, A, E);
 }
 
-/// Handle the [[intel::doublepump]] and [[intel::singlepump]]
-/// attributes.
-template <typename AttrType>
-static void handleIntelFPGAPumpAttr(Sema &S, Decl *D, const ParsedAttr &A) {
-  checkForDuplicateAttribute<AttrType>(S, D, A);
+/// Handle the [[intel::singlepump]] attribute.
+static void handleSYCLIntelFPGASinglePumpAttr(Sema &S, Decl *D,
+                                              const ParsedAttr &AL) {
+  checkForDuplicateAttribute<IntelFPGASinglePumpAttr>(S, D, AL);
+
+  // If the declaration does not have an [[intel::fpga_memory]]
+  // attribute, this creates one as an implicit attribute.
+  if (!D->hasAttr<IntelFPGAMemoryAttr>())
+    D->addAttr(IntelFPGAMemoryAttr::CreateImplicit(
+        S.Context, IntelFPGAMemoryAttr::Default));
+
+  D->addAttr(::new (S.Context) IntelFPGASinglePumpAttr(S.Context, AL));
```

Comment:
Just for my education - Is there a reason to do this instead of calling `handleSimpleAttribute`

Reply:
Thanks @elizabethandrews for the reviews.

>>Is there a reason to do this instead of calling handleSimpleAttribute

Technically there is no difference here as handleSimpleAttribute() function returns the attribute itself. I just tried to skip extra step (not to use another handler function) and be consistent with stall_enable or nofusion attribute which also takes no argument (https://github.com/intel/llvm/pull/5053/)

```
template <typename AttrType>
static void handleSimpleAttribute(Sema &S, Decl *D,
                                  const AttributeCommonInfo &CI) {
  D->addAttr(::new (S.Context) AttrType(S.Context, CI));
}
```